src/node/stage/approved.rs

```diff
@@ -568,19 +570,61 @@ impl Approved {
         }
     }
 
-    pub fn handle_genesis_update(
+    pub fn handle_elders_update(&mut self, core: &mut Core, payload: EldersUpdate) -> Result<()> {
+        info!("Received {:?}", payload);
+
+        core.msg_filter.reset();
+
+        self.shared_state = SharedState::new(payload.elders_info);
+        self.section_keys_provider = SectionKeysProvider::new(None);
+        self.reset_parsec(core, payload.parsec_version)
+    }
+
+    pub fn handle_promote(
         &mut self,
         core: &mut Core,
-        genesis_prefix_info: GenesisPrefixInfo,
-        section_key: bls::PublicKey,
+        shared_state: SharedState,
+        parsec_version: u64,
     ) -> Result<()> {
-        info!("Received GenesisUpdate: {:?}", genesis_prefix_info);
+        if !shared_state.our_info().elders.contains_key(core.name()) {
+            debug!("ignore Promote - not actually promoted");
+            return Ok(());
+        }
 
         core.msg_filter.reset();
 
-        self.shared_state = SharedState::new(genesis_prefix_info.elders_info, section_key);
-        self.section_keys_provider = SectionKeysProvider::new(None);
-        self.reset_parsec(core, genesis_prefix_info.parsec_version)
+        // TODO: verify `shared_state` !
+        self.shared_state.update(shared_state)?;
+
+        self.reset_parsec(core, parsec_version)?;
+        self.gossip_timer_token = Some(core.timer.schedule(self.consensus_engine.gossip_period()));
+
+        info!("Promoted");
+
+        match self
+            .section_keys_provider
+            .finalise_dkg(core.name(), self.shared_state.our_info())
+        {
+            Ok(()) => (),
+            Err(RoutingError::InvalidElderDkgResult) => {
+                // Ignore `InvalidElderDkgResult` because it just means the DKG hasn't completed
+                // for us yet.
+                // TODO: check that we have an ongoing DKG instance corresponding to the latest
+                // EldersInfo.
+            }
+            Err(error) => return Err(error),
+        }
+
+        self.send_elders_update(core)?;
+        self.send_send_neighbour_info()?;
```

Comment:
I think the idea was that we would only send `NeighbourInfo` to other section if we detect their knowledge of us is outdated which happens when we receive a message from them. Why are we doing it proactively here now?

Reply:
This is mainly for the split case. During split, especially for the testing setup, there will be no message flowing from one new-sub-section to the other. Hence we will have to send NeighbourInfo to other section actively.

Reply:
Do you mean SendSibling as opposed to SendNeighbor in this case?

Reply:
Yes. However, didn't want to introduce extra SendSibling message/variant, so used the `neighbour` stuff at that moment.

Reply:
Yes I see, but I mean should that not be renamed as currently it is specific only to Siblings and we don't have a neighbor as a concept unless/until we have to again have neighbors when the network is several millinois of ndoes or whatever.

Reply:
> During split, especially for the testing setup, there will be no message flowing from one new-sub-section to the other.

This sounds wrong. So you are basically adjusting the production code to accommodate the tests, instead of the other way around. In the real network there will always be some messages flowing all the time, so we should not need this. In the tests, we should simulate those message instead. And we actually already do, there is a `update_neighbours_and_poll` function in the test utils for this very purpose.

Reply:
> as currently it is specific only to Siblings 

It's not specific to siblings, as far as I can see. Siblings get updated about each other immediately, because they process the `SectionInfo` vote of both subsections. It's the non-siblings that we need to update. Although my claim is that we actually don't as it happens automatically as messages are being sent over the network.

> and we don't have a neighbor as a concept

We still do. That change hasn't been implemented yet.

Reply:
> Although my claim is that we actually don't as it happens automatically as messages are being sent over the network.
I agree completely with this. 
>We still do. That change hasn't been implemented yet.
Yes cheers I was meaning as a design concept, but yea we need to remove that code.

Reply:
>there is a update_neighbours_and_poll function in the test utils for this very purpose

Reply:
The newly promoted nodes won't vote, so sometimes during split there will be not enough votes.

Reply:
Yes, but that's a different issue - it's about the knowledge update. This issue here is about us sending the `NeighbourInfo` proactively which we shouldn't do.

Reply:
So the resolution of this is that the `send_send_neighbour_info` call should be removed.

Reply:
addressed by PR https://github.com/maidsafe/routing/pull/2170